In [ ]:
import matplotlib.pyplot as plt
from tensorflow import keras

from keras.datasets.imdb import load_data
from keras.optimizers import Adam
from keras.utils import pad_sequences
from keras.models import Sequential
from keras.layers import Embedding, Bidirectional, LSTM, Dense, Dropout

In [ ]:
max_features = 10000
maxlen = 500
batch_size = 32

In [ ]:
(x_train, y_train), (x_test, y_test) = load_data(num_words=max_features)

In [ ]:
x_train = pad_sequences(x_train, maxlen=maxlen)
x_test = pad_sequences(x_test, maxlen=maxlen)

In [ ]:
drnn_model = Sequential()
drnn_model.add(Embedding(max_features, 32))
drnn_model.add(Dropout(0.2))
drnn_model.add(Bidirectional(LSTM(32, return_sequences=True)))
drnn_model.add(Bidirectional(LSTM(32)))
drnn_model.add(Dense(256, activation = "relu"))
drnn_model.add(Dropout(0.2))
drnn_model.add(Dense(1, activation='sigmoid'))

In [ ]:
drnn_model.compile(optimizer=Adam(1e-4), loss='binary_crossentropy', metrics=['acc'])

In [ ]:
drnn_history = drnn_model.fit(x_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

In [8]:
test_loss, test_acc = drnn_model.evaluate(x_test, y_test)
print('Test Accuracy: {}'.format(test_acc))

782/782 [==============================] - 128s 163ms/step - loss: 0.3315 - acc: 0.8749
Test Accuracy: 0.8749200105667114


# Conclusion

In this homework assignment, we created 4 different recurrent neural networks:  

RNN:  
 * Time of learning: 180s 
 * Accuracy: 86.6%

LSTM:  
 * Time of learning: 1920s
 * Accuracy: 86.7%

BRN:  
 * Time of learning: 2000s
 * Accuracy: 86.8%

DRN:  
 * Time of learning: 1500s
 * Accuracy: 87.5%